### PR TITLE
Basic implementation of RenderPass based architecture

### PR DESCRIPTION
**Pull request by mvaligursky**
This is a first step in the implementation of https://github.com/playcanvas/engine/issues/4271. ForwardRenderer, instead of directly rendering the scene, performs two steps:
1. analyses the scene, and constructs a FrameGraph of rendering passes. Only layers rendered by the cameras are fully split into render passes, other rendering systems (shadows, grab passes, post-effects and similar) are inserted into the frame graph as black boxes. Those will be further broken down into actual render passes in the future.
2. renders the generated render passes

Other changes:
- making Debug.setTrace public, to allow external applications to enable log tracing
- added few trace ids for debug logging related to render passes

example of the trace output for post-effects engine example
Screenshot 2022-05-25 at 11 47 48

and here are the frame passes from the same example when postprocessing is disabled - this results in a smaller number of render passes, which in the future PRs will translate to a smaller number of multi-sampled resolves:
Screenshot 2022-05-25 at 11 49 41

### CODE REVIEW

File: src/core/debug.js [hunk 1]
```diff
@@ -24,7 +44,11 @@ class Debug {
     /**
      * Enable or disable trace channel
      *
-     * @param {string} channel - Name of the trace channel
+     * @param {string} channel - Name of the trace channel. Can be
```
Comment (willeastcott) [inline src/core/debug.js:47]:
```
     * @param {string} channel - Name of the trace channel. Can be:
```

File: src/scene/render-pass.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * A render pass represents a node in the frame graph, and encapsulates a system which
+ * renders to a render target using an execution callback.
+ *
+ * @ignore
+ */
+class RenderPass {
+    /** @type {string} */
+    name;
+
+    constructor(renderTarget, execute) {
```
Comment (willeastcott) [inline src/scene/render-pass.js:13]: Might be a good idea to add JSDoc comments for this constructor. `execute` in particular is unexplained here.

File: src/scene/renderer/forward-renderer.js [hunk 3]
```diff
@@ -81,6 +86,9 @@ const _tempMaterialSet = new Set();
  * The forward renderer renders {@link Scene}s.
  */
 class ForwardRenderer {
+    /** @type {boolean} */
```
Comment (willeastcott) [inline src/scene/renderer/forward-renderer.js:89]: Shouldn't this have `@ignore`?

**Review by willeastcott — APPROVED** [review]

File: src/scene/render-pass.js (hunk 2 above)
Comment (mvaligursky) [reply src/scene/render-pass.js:13]: yep, done

File: scripts/utils/cubemap-renderer.js [hunk 4]
```diff
@@ -70,6 +70,7 @@ CubemapRenderer.prototype.initialize = function () {
 
         // render target, connected to cubemap texture face
         var renderTarget = new pc.RenderTarget({
+            name: `CubemapRenderer-Face${i}`,
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply scripts/utils/cubemap-renderer.js:73]: great point, on it

**Pull request merged**